src/layers/embeddings.jl

```diff
@@ -1,27 +1,35 @@
 """
-    PatchEmbedding(patch_size)
-    PatchEmbedding(patch_height, patch_width)
+    PatchEmbedding(imsize::NTuple{2} = (224, 224); inchannels = 3, patch_size = (16, 16), 
+                   embedplanes = 768, norm_layer = identity, flatten = true)
 
-Patch embedding layer used by many vision transformer-like models to split the input image into patches.
+Patch embedding layer used by many vision transformer-like models to split the input image into 
+patches.
+
+# Arguments:
+- `imsize`: the size of the input image
+- `inchannels`: the number of channels in the input image
+- `patch_size`: the size of the patches
+- `embedplanes`: the number of channels in the embedding
+- `norm_layer`: the normalization layer
+- `flatten`: whether to flatten the input after the embedding
 """
-struct PatchEmbedding
-  patch_height::Int
-  patch_width::Int
-end
+function PatchEmbedding(imsize::NTuple{2} = (224, 224); inchannels = 3, patch_size = (16, 16), 
+                        embedplanes = 768, norm_layer = identity, flatten = true)
 
-PatchEmbedding(patch_size) = PatchEmbedding(patch_size, patch_size)
+  im_height, im_width = imsize
+  patch_height, patch_width = patch_size
 
-function (p::PatchEmbedding)(x)
-  h, w, c, n = size(x)
-  hp, wp = h ÷ p.patch_height, w ÷ p.patch_width
-  xpatch = reshape(x, hp, p.patch_height, wp, p.patch_width, c, n)
+  norm_layer = typeof(norm_layer) != typeof(identity) ? norm_layer(embedplanes) : identity
 
-  return reshape(permutedims(xpatch, (1, 3, 5, 2, 4, 6)), p.patch_height * p.patch_width * c, 
-                 hp * wp, n)
+  @assert (im_height % patch_height == 0) && (im_width % patch_width == 0)
+  "Image dimensions must be divisible by the patch size."
+  
+  return Chain(Conv(patch_size, inchannels => embedplanes; stride = patch_size),
+               flatten ? x -> permutedims(reshape(x, (:, size(x, 3), size(x, 4))), (2, 1, 3)) 
+                       : identity,
+               norm_layer)
```

Comment:
This is quite different from before. The patches now have a learnable weight associated with them? That seems limiting to me. I would separate the learnable portion + normalization from the simple breaking up of an image into patches.

Reply:
Um, so this is the part that changed - the initial implementation wasn't a `PatchEmbedding`, it was just a Patching layer with a dense layer afterwards. But when I was seeing some more implementations, I realised that this learnable projection goes along with the patches every single time (for context, the timm implementations - from which the official repos of most recent papers borrow layers, use `PatchEmbed` in this fashion) - so it made sense to combine them in a format that's much more recognisable and also cleaner

Reply:
Okay that makes sense. I suppose this should be more performant then. What about the normalization? It doesn't seem to be always used, and I see it kind of similar to batch norm. It's common to have conv + bn but we still keep the layers separate and provide a helper to `Chain` them. Maybe something like having the `PatchEmbedding` and `PatchNormalization` (which has the flatten + norm)?

Another note is that now this layer doesn't actually need to be a separate type...so naming issue again.

Reply:
The naming issues probably need to be decided on in a common format (because this is quite the common layer and camel-casing was my first thought 😅) but I'll default to the overall consensus. For the norm - `LayerNorm` has been used quite often, so that's probably one of the common choices, but this choice was written less from a current standpoint and more from the POV of people writing future models using these building blocks I reckon